File: platform-ui/src/app/portal/admin/services/real-api/dashboard.service.ts
import { Injectable } from '@angular/core';
import { map, Observable } from 'rxjs';
import {
  BorrowerMetrics,
  CategoryMetrics,
  DashboardBorrowesMetrics,
  DashboardBorrowesOverTimeData,
  DashboardsAdminApiService,
} from '../../../../api-client';
import {
  DashboardService,
  UIBorrowerMetrics,
  UICategoryMetrics,
  UIDashboardBorrowesMetrics,
  UIDashboardBorrowesOverTimeData,
  UIItemMetrics,
} from '../dashboard.service';

@Injectable({
  providedIn: 'root',
})
export class ApiDashboardService implements DashboardService {
  constructor(private dashboardsAdminApiService: DashboardsAdminApiService) {}

  getDashboardData(): Observable<UIDashboardBorrowesOverTimeData> {
    return this.dashboardsAdminApiService.getDashboardData().pipe(
      map((data: DashboardBorrowesOverTimeData) => ({
        labelsX: data.labelsX,
        axisYLabels: data.axisYLabels,
        data: data.data.map((point) => [point.x, point.y] as [number, number]),
      })),
    );
  }

  getUserCount(): Observable<number> {
    return this.dashboardsAdminApiService.getUserCount();
  }

  getItemCount(): Observable<number> {
    return this.dashboardsAdminApiService.getItemCount();
  }

  getLibraryCount(): Observable<number> {
    return this.dashboardsAdminApiService.getLibraryCount();
  }

  getDashboardMetrics(): Observable<UIDashboardBorrowesMetrics> {
    return this.dashboardsAdminApiService.getDashboardMetrics().pipe(
      map((data: DashboardBorrowesMetrics) => ({
        totalBorrows: data.metrics.totalBorrows ?? 0,
        totalReservations: data.metrics.totalReservations ?? 0,
        itemsOnLoan: data.metrics.itemsOnLoan ?? 0,
        avgLoanDuration: data.metrics.avgLoanDuration + ' days',
        returnTimeliness: data.metrics.returnTimeliness + '%',
      })),
    );
  }

  getTopBorrowers(): Observable<UIBorrowerMetrics[]> {
    return this.dashboardsAdminApiService.getTopBorrowers().pipe(
      map((data: BorrowerMetrics[]) =>
        data.map((borrower) => ({
          username: borrower.username,
          totalBorrows: borrower.totalBorrows,
        })),
      ),
    );
  }

  getTopItems(): Observable<UIItemMetrics[]> {
    return this.dashboardsAdminApiService.getTopItems();
  }

  getTopCategories(): Observable<UICategoryMetrics[]> {
    return this.dashboardsAdminApiService.getTopCategories().pipe(
      map((data: CategoryMetrics[]) =>
        data.map((category) => ({
          category: category.category,
          totalBorrows: category.totalBorrows,
          numberOfItems: category.borrowedItems ?? 0,
        })),
      ),
    );
  }
}

File: platform-api/src/main/java/dev/theopenshelf/platform/services/DashboardService.java
package dev.theopenshelf.platform.services;

import java.util.List;
import java.util.stream.Collectors;

import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import dev.theopenshelf.platform.model.BorrowerMetrics;
import dev.theopenshelf.platform.model.CategoryMetrics;
import dev.theopenshelf.platform.model.DashboardBorrowesMetrics;
import dev.theopenshelf.platform.model.DashboardBorrowesMetricsMetrics;
import dev.theopenshelf.platform.model.DashboardBorrowesOverTimeData;
import dev.theopenshelf.platform.model.ItemMetrics;
import dev.theopenshelf.platform.repositories.BorrowRecordRepository;
import dev.theopenshelf.platform.repositories.CategoryRepository;
import dev.theopenshelf.platform.repositories.CommunityRepository;
import dev.theopenshelf.platform.repositories.ItemsRepository;
import dev.theopenshelf.platform.repositories.UsersRepository;
import lombok.RequiredArgsConstructor;
import reactor.core.publisher.Mono;

@Service
@RequiredArgsConstructor
public class DashboardService {
    private final BorrowRecordRepository borrowRecordRepository;
    private final CategoryRepository categoryRepository;
    private final CommunityRepository communityRepository;
    private final ItemsRepository itemRepository;
    private final UsersRepository userRepository;

    public Mono<DashboardBorrowesOverTimeData> getDashboardData() {
        // TODO: Implement this method to return borrow data over time
        return Mono.just(DashboardBorrowesOverTimeData.builder().build());
    }

    public Mono<DashboardBorrowesMetrics> getDashboardMetrics() {
        return Mono.fromCallable(() -> {
            long activeBorrows = borrowRecordRepository.countActiveBorrows();
            long activeReservations = borrowRecordRepository.countActiveReservations();
            long totalBorrows = borrowRecordRepository.countTotalBorrows();

            return DashboardBorrowesMetrics.builder()
                    .metrics(DashboardBorrowesMetricsMetrics.builder()
                            .itemsOnLoan((int) activeBorrows)
                            .totalReservations((int) activeReservations)
                            .totalBorrows((int) totalBorrows)
                            .build())
                    .build();
        });
    }

    public Mono<List<BorrowerMetrics>> getTopBorrowers() {
        return Mono
                .fromCallable(() -> borrowRecordRepository.findTopBorrowers(PageRequest.of(0, 5)).getContent().stream()
                        .map(row -> BorrowerMetrics.builder()
                                .username((String) row[0])
                                .totalBorrows(((Number) row[1]).intValue())
                                .build())
                        .collect(Collectors.toList()));
    }

    public Mono<List<CategoryMetrics>> getTopCategories() {
        return Mono.fromCallable(() -> categoryRepository.findTopCategories(PageRequest.of(0, 5)).getContent().stream()
                .map(row -> CategoryMetrics.builder()
                        .category((String) row[0])
                        .icon((String) row[1])
                        .totalBorrows(((Number) row[2]).intValue())
                        .borrowedItems(((Number) row[3]).intValue())
                        .build())
                .collect(Collectors.toList()));
    }

    public Mono<List<ItemMetrics>> getTopItems() {
        return Mono.fromCallable(() -> itemRepository.findTopItems(PageRequest.of(0, 5)).getContent().stream()
                .map(row -> ItemMetrics.builder()
                        .item((String) row[0])
                        .totalBorrows(((Number) row[1]).intValue())
                        .build())
                .collect(Collectors.toList()));
    }

    public Mono<Integer> getUserCount() {
        return Mono.fromCallable(() -> (int) userRepository.count());
    }

    public Mono<Integer> getItemCount() {
        return Mono.fromCallable(() -> (int) itemRepository.count());
    }

    public Mono<Integer> getLibraryCount() {
        return Mono.fromCallable(() -> (int) communityRepository.count());
    }
}